docs/src/pages/ThemingSpecification.mdx

```diff
@@ -0,0 +1,185 @@
+
+In the rest of the specification, there will be some proposals of how the themes, mappings and utilities can be structured.
+
+#### CSS stylesheets usage
+
+Users should be able to define mapping from props to classNames per component level.
+
+This is inspired by the [React Semantic UI's className builders](https://github.com/Semantic-Org/Semantic-UI-React/blob/master/src/lib/classNameBuilders.js)
+
+Here is an example of how the mappings for a FluentButton component can look like:
+
+```
+const fluentButtonMappings = props => {
+  const { size, circular, attached, corner } = props
+  return cx(
+    size, // the value (small, medium, big)
+    useKeyOnly(circular, 'circular'),
+    useKeyOrValueAndKey(attached, 'attached'), // true or the value of the attached prop
+    useValueAndKey(corner, 'corner'), // corner="left" will create the classNames `corner left`
+  )
+}
+
+const FluentButton = compose(Button, {
+  displayName: 'FluentButton',
+  mapPropsToClasses: fluentButtonMappings,
+})
+```
+
+#### CSSinJS usage
+
+If users decide to use CSSinJS we have to propose some structure of how they can define all things required for theming their components.
+
+I will describe one proposal of how the theme object can look like:
+
+##### Global tokens
+
+These are tokens that can be reused in many components, things like density, font sizes, font families, colors, shadows etc. Having these global tokens, can help designers to have more consistent look and feel on the components which are part of one application/theme.
+
+The global tokens can be defined as a plain javascript object. We can offer some utilities for generating color palettes if there is an agreement of how those should be generated, but anyway, the global tokens should be defined by the users in any format they would like to structure them, and our framework should not depend on something that is defined (or not defined) there.
+
+##### Component tokens
+
+These are tokens which are specific for a component. They can use some of the global tokens for their values.
+
+**Open for a debate here**
+
+We have the components tokens in Teams theme and those were shown as a good abstraction. On the other hand, people overused them for adding inline overrides, because these tokens can be overridden on any instance of a component. So, maybe the best option will be to restrict the ways of where these tokens can be defined (only in the theme and not on instance level), so we can avoid similar misusages in future.
+
+##### Component styles
+
+The component styles should be functions that return CSSinJS definition that is compatible with the CSSinJS engine used. The input parameters of these functions are:
+
+- **design terms** - in the react implementation, these are basically subset of the component's properties. It is important to note here that for better performance we need to restrict these to be primitive values like: booleans, strings, numbers, so we can easily hash based on them.
```

Comment:
> subset of the component's properties

Could you add some examples of what properties these are?

Reply:
Maybe the description is not the best. For the React components we can have many props. For example for a FluentButton we may have: onClick, content, icon, disabled, primary, circular etc. Out of these only subset of them are design terms: disabled, primary and circular.